### PR TITLE
test/boost/view_schema_test: Improve test_view_update_generating_writetime

### PR DESCRIPTION
**Pull request by dawmd**
In this PR, we improve various aspects of the test:

* increase obtained information whenever any test case fails,
* split test cases,
* elaborate on the semantics of generating view updates
  and what exactly we check and why.

Backport: not needed, this is an enhancement.

### CODE REVIEW

**Comment by dawmd** [comment]
Ad commit 2 (changing comments):

I don't claim that my explanation of the semantics of generating view updates is true. However, it's the only explanation that I've been able to think of for the test to make sense.

According to the comments scattered across the repository, it seems like virtual columns are only a thing when the primary keys in the base table and a materialised view are the same. However, if that were to be true, the test case when we set and update `f` should only generate a view update for mv1, but not for mv2. That's not what happens and that's not what is expected. What *is* expected is that we generate view updates for both views.

But if virtual columns existed for *each* unselected column for *any* view, then the whole test would immediately make sense. Please help me find the truth and I'll update the description.

File: test/boost/view_schema_test.cc [hunk 1]
```diff
@@ -2993,6 +2993,16 @@ SEASTAR_TEST_CASE(test_base_column_in_view_pk_complex_timestamp) {
...
+namespace std {
+
+// Needed by `test_view_update_generating_writetime` below.
+ostream& operator<<(ostream& os, std::tuple<int, int, int> tuple) {
+    os << seastar::format("{}", tuple);
+    return os;
+}
+
+} // namespace std
+
```
Comment (wmitros) [inline test/boost/view_schema_test.cc:3005]: If we're improving readability, maybe we can also print what each value means (I see all tuples below have the same meaning, we could use a struct instead)

**Review by wmitros — COMMENTED** [review]
The only change in how the test executes should be the ttl change and that should be fine.
I added a nitpick about the logging change.

I read the comments and they look good, but maybe @nyh should look at them as I not that familiar with how we handle timestamps and virtual columns in views.

Regarding virtual columns in views, from what I'm seeing it looks like we really don't create them for views with non-base-key columns in their keys. The fact that we generate updates for them anyway (as in the 'f' case) sounds like a bug - and I might have found where:
* in `view_updates::generate_update` we reach `has_old_row` && `has_new_row` && `same_row`, so we enter 'update_entry'
* in `view_updates::update_entry` we call `can_skip_view_updates` which should probably prevent us from sending the view update
* in `view_updates::can_skip_view_updates` we no longer check whether we have non-base-key columns in view key, but we check whether some new base column value appeared with a nonexpiring marker - this happens in the first statement in the `f` case, so we send the update anyway, even though the view doesn't have virtual columns.

**Comment by nyh** [comment]
>     * prevent a possible unrelated failure related to an expiring row
>       if the conditions in the last test case fail to be satisfied 13 times
>       in a row.

Is this related to, or perhaps meant to solve, https://github.com/scylladb/scylladb/issues/20889 ? I guess not?

File: test/boost/view_schema_test.cc [hunk 2]
```diff
@@ -3104,7 +3104,7 @@ SEASTAR_TEST_CASE(test_view_update_generating_writetime) {
         // The same scenario as before, but this time the second statement specifies TTL,
         // so updating the cell WILL generate a view update for each MV too.
         e.execute_cql("UPDATE t USING TIMESTAMP 7 SET g=40 WHERE k=1 AND c=1;").get();
-        e.execute_cql("UPDATE t USING TTL 10 AND TIMESTAMP 8 SET g=40 WHERE k=1 AND c=1;").get();
+        e.execute_cql("UPDATE t USING TTL 300 AND TIMESTAMP 8 SET g=40 WHERE k=1 AND c=1;").get();
```
Comment (nyh) [inline test/boost/view_schema_test.cc:3107]: Maybe eventually() shouldn't have defaulted to 263 seconds (!), but as long as it does I agree the TTL (which we don't intend to delete the item, just to exist) should be higher. So good.

**Review by nyh — COMMENTED** [review]
Looks good to me, but I suggest (unless if there's a reason not to) to split the two-updates-and-then-check-counts-once sequence to make it easier to understand and also verify the counts between the two updates are really as you explain in the comment.

**Comment by dawmd** [comment]
> > ```
> > * prevent a possible unrelated failure related to an expiring row
> >   if the conditions in the last test case fail to be satisfied 13 times
> >   in a row.
> > ```
> 
> Is this related to, or perhaps meant to solve, #20889 ? I guess not?

It's related (as in the problem this commit solves manifests in the logs of the failed tests in #20889), but it doesn't solve it unfortunately

File: test/boost/view_schema_test.cc (hunk 1 above)
Comment (dawmd) [reply test/boost/view_schema_test.cc:3005]: That's a good idea, thanks!

**Comment by piodul** [comment]
> It's related (as in the problem this commit solves manifests in the logs of the failed tests in #20889), but it doesn't solve it unfortunately

If it's related then it's worth putting `Refs: scylladb/scylladb#20889` into the commit / PR description.

**Comment by dawmd** [comment]
> If it's related then it's worth putting `Refs: scylladb/scylladb#20889` into the commit / PR description.

Sorry, I should've worded it a bit better. The problem appears in the logs in that issue, but it's not really related to the actual problem there. Failing `eventually()` 14 times in a row (*no matter why*) will lead to the same test failure. If it's worth adding the line to the commit, I'll do that.

**Comment by nyh** [comment]
> > If it's related then it's worth putting `Refs: scylladb/scylladb#20889` into the commit / PR description.
> 
> Sorry, I should've worded it a bit better. The problem appears in the logs in that issue, but it's not really related to the actual problem there. Failing `eventually()` 14 times in a row (_no matter why_) will lead to the same test failure. If it's worth adding the line to the commit, I'll do that.

No, if it's a different problem, you probably don't need to mention the issue.  I wondered if maybe you suspected that "failing 14 times in a row" was the problem we originally saw and just forgot to mention it - but I do agree that it probably isn't the same issue, and what you fixed just a different hypothetical problem that never actually happens (which doesn't mean, of course, you can't premptively fix it).

**Comment by dawmd** [comment]
Version 3:
* Fixed a bug in the logic responsible for generating view updates. Updated the test (because it assumed the wrong semantics).
* Ditched the commit changing the TTL (because it's handled by scylladb/scylladb#21632, which has been merged).
* Split the test cases for better readability.
* Applied Wojciech's suggestion to create a custom type for view update counters.
* Modified the comments in the code a bit and elaborated on the semantics of virtual columns, etc.

**Comment by nyh** [comment]
Once the PR includes a change to the algorithm (db/view.cc), not just the test, I think its name and the focus of the commit message needs to change.

File: test/boost/view_schema_test.cc [hunk 3]
```diff
@@ -3011,6 +3011,21 @@ struct update_counter {
 };
 
 SEASTAR_TEST_CASE(test_view_update_generating_writetime) {
+    // In this test, we verify that generating view updates abides by the intended algorithm.
```
Comment (nyh) [inline test/boost/view_schema_test.cc:3014]: This sentence is very general, the whole file is about checking that view updates abide by the intended algorithm :-)
See below my proposal how the raison d'etre of this test can be explained.

File: test/boost/view_schema_test.cc [hunk 4]
```diff
@@ -3011,6 +3011,21 @@ struct update_counter {
...
+    //        it will have virtual columns corresponding to unselected columns from the base table.
+    //        Creating a value in such a column (in the base table) will generate a view update
+    //        to the MV. However, updating it will not generate an update UNLESS it changes
+    //        the cell's TTL.
+    // * mv2: its primary key consists of the columns from the base table's primary key and one
+    //        regular column. Because of that, the MV will NOT have any virtual columns corresponding
+    //        to the unselected columns from the base table. As a result, no view updates will be
+    //        generated for unselected columns as a result.
+    //
+    // We also verify that `WRITETIME` works correctly.
```
Comment (nyh) [inline test/boost/view_schema_test.cc:3027]: This sentence is the main goal of this test and I think it should come first. However, it's not about testing the `WRITETIME` function like your phrasing might suggest to readers. Rather - it's about the timestamps written to the view table being set to the correct value (same as in the base table). But, we check **only** the timestamps that are actually readable (using `WRITETIME()`) in the view - in other words, only the timestamps of selected columns. The timestamp of columns not readable with `WRITETIME()` - namely virtual columns (created in some cases for unselected columns) do not matter, so Scylla can - and does - avoid updating them, and may skip some view updates altogether.  The second goal of this test is to verify that view updates are indeed skipped when their only function would have been to update a non-readable timestamp.

**Review by nyh — CHANGES_REQUESTED** [review]
Looks mostly good to me (thanks!), but I do have a few more requests and suggestions.

File: test/boost/view_schema_test.cc (hunk 4 above)
Comment (dawmd) [reply test/boost/view_schema_test.cc:3027]: I think I get it now, thank you for your explanation!

**Comment by dawmd** [comment]
Version 4:
* Split the PR into two (the other one: scylladb/scylladb#21653).
* Improved the test's description.

**Comment by dawmd** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13300/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13300/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/view_schema_test
> #### Failed Tests (1/62):
> 
>     * [test_view_update_generating_writetime](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13300/testReport/junit/boost/view_schema_test/test_view_update_generating_writetime) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_update_generating_writetime)
> 
> 
> #### Build Details:
> 
>     * Duration: 48 min
> 
>     * Builder: spider8.cloudius-systems.com

The failure occurred because scylladb/scylladb#21653 is still pending. We'll most likely need to rebase this PR once is on the master, so let's not update this one and let's just wait.

**Comment by dawmd** [comment]
Version 5:
* Rebased on top of the current master branch.

File: test/boost/view_schema_test.cc [hunk 5]
```diff
@@ -3029,37 +3046,45 @@ SEASTAR_TEST_CASE(test_view_update_generating_writetime) {
         eventually([&] {
             msg = e.execute_cql("SELECT WRITETIME(e) FROM t").get();
             assert_that(msg).is_rows().with_row({long_type->decompose(int64_t(1))});
-            BOOST_REQUIRE_EQUAL(total_t_view_updates(), 1);
-            BOOST_REQUIRE_EQUAL(total_mv1_updates(), 1);
-            BOOST_REQUIRE_EQUAL(total_mv2_updates(), 0);
+
+            const update_counter results{total_mv1_updates(), total_mv2_updates(), total_t_view_updates()};
```
[the remark this answers is not in the document]
Comment (nyh) [reply test/boost/view_schema_test.cc:3073]: Ok, I don't insist. The only reason why maybe it makes sense to keep that total test is that it checks completely different counters - so maybe can expose a bug in those counters or something.

**Review by nyh — APPROVED** [review]

**Pull request merged**